### DISCUSSION TITLE
February Developer Meeting

### DISCUSSION THREAD

**Discussion by TheDuckCow** [Moo-Ack-Productions/MCprep#540] [General Discussions]
Starting this discussion thread as a more public and fitting space to propose some ideas, wishlists, and priorities for upcoming MCprep development.

I'll share more of my own thoughts in a follow up reply here tonight, but as a starter I'll share these couple links. Let's not treat these milestone as set in stone, but rather a starting point to consider for the roadmap. The output of the meeting tomorrow will be to update them based on the discussion.

- [Draft v3.5.3](https://github.com/Moo-Ack-Productions/MCprep/milestone/15)
- [Draft v3.6.0](https://github.com/Moo-Ack-Productions/MCprep/milestone/13)
- [Draft v4.0.0](https://github.com/Moo-Ack-Productions/MCprep/milestone/14)

**Comment by TheDuckCow** [comment]
And so, to share my own thoughts - treat these as opinions to be discussed tomorrow:

- General philosophy: Better for us to think smaller, with a more reliable output. So consider that with the current-state milestones to be reduced or split up further
- v3.5.3 is of course a good near-term canddiate of misc fixes and QoL improvements that are within reach.
  - However, it's also turned into a bit of a laundry list and nice to haves that keep being punted. Classic example: https://github.com/Moo-Ack-Productions/MCprep/issues/83 - nobody in particular has actually asked for this or complained about it (most users only use one scene per project file anyways), it just conceptually seemed like a nice thing to do. Does not mean it should be prioritized if there's not user value, though it could still remain as a good first contribution
  - From this current milestone, 
- v3.6.0 to focus on near-term features, realistic for the next month I'd argue (ie Q1'24)
  - Internationalization is a good primary focus for this version.
- And now, let's chat v4.0:
  - The reason for the major version uptick is because it would represent a major overhaul from the user experience flow.
  - Re-organization to make content more discoverable. Spawners are especially awkward to scroll down to the right one today
  - Maybe asset browser integration (or pack generation) should be more of a priority?
  - This major UI upgrade will also require a refreshed 1-minute tutorial series. TBH that's a tall order for me, but they also probably (falsely) look out of date since they began in the blende 2.9x era. This could be a good motivator for me to prioritize a re-make.
- Given the spirit of small but attainable milestone, we should think about the stepping stones we could make to get to 4.0.
  - Maybe we define a 3.7 and 3.8 which feature more of the core backend updates, without a complete UI workflow mashup which should belong to 4.0
  - Good candidates for that would be the mechanisms for auto loading/extracting textures from user machines, steps towards async asset updates vs mcprep updates. Depending on how complex though, might be better fitting for 4.0 anyways.
- On the direct bridge loading (ie that other big initiative that I've been punting since... well, since [Blender 2.7](https://twitter.com/TheDuckCow/status/1151276654787989505))
  - While I still worry about the maintainability of this feature, I think it may be worthwhile to have as an opt-in and lead with an expectation of no specific support to be given for the feature.
  - The only thing I'm torn about is that _this_ feature, if we could be confident about it, could make for a nice centralization of the workflow - everything could be re-focussed around it and be well integrated (texture packs, areas for meshswap, extending imports for sections of the world).
  - But, right now it looks a bit more like a pipe dream, or at least a more 4.1 or even 5.0 kind of feature.

**Comment by Jaitie018** [comment]
Something that fixes Z-Fighting on imported worlds?

**Reply by StandingPadAnimations** [reply]
This is known behavior with Mineways as Mineways adds double faces for certain renderers. This can be fixed in the export settings for Mineways

**Comment by TheDuckCow** [comment]
Thanks @StandingPadAnimations for joining in, and @Jaitie018 for sharing your topic idea!

**My summary of the dev meeting:**

- Let’s talk first about MCprep 4.0
    - Should be the UI overhaul and new workflow with layering of resource packs.
    - Should also include the separation of asset installs from the plugin; makes updating faster.
    - Would also be nice to deprecate the BSS rigs since they are in some places broken and mostly unsupported. Should be only removed with replacements however, and to the new standardized quality.
- And for 3.6?
    - Primary focus is on translations.
    - Also brought up: Z-fighting issue, need to investigate Mineways settings. Essentially, if we detect the setting that is to be avoided, recommend a re-export to remove it.
    - This is better than the “merge by distance” solution, which can have unintended/random results
    - Should also remove the Cycles Optimizer, StandingPad to create the task for this.
- Chatting more about translations:
    - Need to be able to register the bl_label, via bpy.app.translations. Need to also capture tooltips, but we could use the dynamic properties.
    - We also want to ease the contribution of translations as much as possible, and git/merging/PR creation is a barrier to entry. Looks like [Weblate](https://weblate.org/en/hosting/) supports open source projects for free. Seems it can result in po files that we’d want.
        - We should do this as an experimental after 3.6, since it should fit in with the current setup.
    - Clarifying the workflow: Generating the po file is always automatic (ie a command), but pre-existing translated po files are patched in, so work is not lost. Might require using a po editor, but hopefully weblate could do this anyways.
    - [Per blender guidelines](https://developer.blender.org/docs/handbook/translating/translator_guide/#translating-non-official-add-ons), seems like there’s a way to load Po files directly. They get patched into the equiv of the translation dictionary.
    - Main conundrum: need to create / auto generate a py file with the bl_idname’s, so it could be picked up/created as a part of the po template. Basiclaly grepping for each of the translation keys.
    - How is blender implementing `is_fuzzy` (reference on [this page](https://docs.blender.org/api/current/bpy.app.translations.html))? Should look into this, in case helpful (or to be avoided).
- Aim to create more releases for 3.7 and 3.8 etc with one primary feature focus each, with smaller timeframes.
- Discussion about defaults in Mineways
    - Did already change from atlas to individual textures as a default.
    - Related thread: https://github.com/erich666/Mineways/issues/99
- Talking about hackathon ideas ([view here](https://github.com/orgs/Moo-Ack-Productions/projects/2/views/3); private currently, may be moved to another board)
    - Could be good for experimental features to get them in place.
    - Could be a safer enviornment for experimentally, wouldn’t necessarily be merged to dev. Those that complete “successfully” could be merged, but still gated under a warning experimental flag.
    - Geo nodes replacement of the dupliverts idea
    - Idea of extracting and using only the first frame of an animated texture as a default instead of not addressing tiled images at all when found. (after block spawning etc). Also the fix for multiple columns.
